fix: show `-` for empty data in run table

keita-determined opened the pull request:
<!---
## PR TITLE (Commit Body)
When squash-merging, GitHub will use this as the commit message.
Check the "Example Commit Body" for conventional commit semantics.
-->
## Ticket
<!---
A reference to the Jira ticket or Github issue. e.g. "[DET-1234]" or #123.
-->
[ET-80]
Related PR: https://github.com/determined-ai/hew/pull/135

## Description
<!---
A description of the PR. For breaking changes, lead with "BREAKING CHANGE:".
--->
Empty entries in the run table should contain a dash


## Test Plan
<!---
Describe the scenarios in which you've tested your change, with screenshots as
appropriate. Reviewers may ask questions about this test plan to ensure adequate
coverage of changes.
-->
- If the cell value is empty (null value), show the dash `-` in the run table


## Checklist

- [x] Changes have been manually QA'd
- [ ] New features have been approved by the corresponding PM
- [ ] User-facing API changes have the "User-facing API Change" label
- [ ] Release notes have been added as a separate file under `docs/release-notes/`
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses have been included for new code which was copied and/or modified from any external code

<!---
Example Commit Body:
docs: tweak recommended "pip install" usage [DET-123]

Specifically, this title should contain a type and a description
of the change being made:

User-facing change types:

- docs: docs-only change
- feat: new user-facing feature
- fix: bug fix
- perf: performance improvement

Internal change types:

- build: build system change (anything in a `Makefile`, mostly)
- chore: any internal change not covered by another type
- ci: anything that touches `.circleci`
- refactor: internal refactor
- style: style change
- test: new tests

See https://www.conventionalcommits.org/en/v1.0.0/ for background.

The first line should also:

- be at most 89 characters long
- contain a description that is at most 72 characters long
- not end with sentence-ending punctuation
- start (after the type) with a lowercase imperative ("add", "fix")
-->


[ET-80]: https://hpe-aiatscale.atlassian.net/browse/ET-80?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

EmilyBonar reviewed (changes requested):
![Screenshot 2024-08-30 at 09-37-56 Uncategorized Runs - Determined](https://github.com/user-attachments/assets/b29a49a5-6afa-43e4-90dc-3a486f33f373)
- Forked column doesn't have dashes
- I think it would look better if the dashes were centered in the column, if possible
- It doesn't make much sense to allow users to right click and copy the "value" of `-` for an empty column. That's confusing behavior and it would be better if the dash was display only.

keita-determined commented:
> Forked column doesn't have dashes

fixed 
Note: ideally, empty `fork` cell should be shown as text, but  it's difficult to change the cell type when the data is empty. Even if there's no fork data, empty fork data, `-`, is shown as a link. The link doesn't go anywhere, and cursor icon is not a hand icon. 

https://github.com/user-attachments/assets/d0cc2d67-b8e9-4ace-9d57-d11a0db068b2


> I think it would look better if the dashes were centered in the column, if possible

it's difficult to make them centered.

> It doesn't make much sense to allow users to right click and copy the "value" of - for an empty column. That's confusing behavior and it would be better if the dash was display only.

fixed. removed copy button if data is empty.

EmilyBonar commented:
> > Forked column doesn't have dashes
> 
> fixed Note: ideally, empty `fork` cell should be shown as text, but it's difficult to change the cell type when the data is empty. Even if there's no fork data, empty fork data, `-`, is shown as a link. The link doesn't go anywhere, and cursor icon is not a hand icon.

> > I think it would look better if the dashes were centered in the column, if possible
> 
> it's difficult to make them centered.

Couldn't you create a `EMPTY_CELL` constant that looked something like:
```
{
      allowOverlay: false,
      data: '-',
      displayData: '-',
      kind: GridCellKind.Text,
      contentAlign: 'center'
}
```
And then you could just the following, which should fix both of the above problems:
```
renderer: (record: FlatRun) => (record.checkpointSize ? 
    {
      allowOverlay: false,
      copyData: humanReadableBytes(record.checkpointSize),
      data: { kind: TEXT_CELL },
      kind: GridCellKind.Custom,
    } : 
    EMPTY_CELL),
```

keita-determined commented:
@EmilyBonar thats a good idea. made some changes. 
not sure if we wanna do `contentAlign: 'center'` though. w&b and some others follows the column alignment. in our case, left aligned is the default. Also, centered dash might be harder to look when column width is long. I think we should ask product people first if we wanna make that move anyways.

EmilyBonar commented on webui/react/src/pages/FlatRuns/columns.ts:
```diff
@@ -99,6 +104,13 @@ function getCellStateFromExperimentState(expState: RunState) {
+  allowOverlay: false,
+  copyData: '-',
+  data: { kind: TEXT_CELL },
+  kind: GridCellKind.Custom,
```
nit: why should this be a custom cell type instead of just text?

keita-determined replied:
That causes type error with `GridCellKind.TEXT`.

EmilyBonar replied:
It's merged now so I guess it's too late, but this doesn't give me a type error:
```
const EMPTY_CELL: GridCell = {
  allowOverlay: false,
  data: '-',
  kind: GridCellKind.Text,
};
```

keita-determined replied:
<img width="972" alt="image" src="https://github.com/user-attachments/assets/c41fc7f3-7a82-4f88-9b82-26fc39f2f40d">

looks like this works. i think we can change it with searches table changes

```ts
const EMPTY_CELL: GridCell = {
  allowOverlay: false,
  data: '-',
  displayData: '-',
  kind: GridCellKind.Text,
} as const;
```

EmilyBonar reviewed (approved):
I feel like an empty column marker being centered is fairly standard, but I suppose it doesn't matter. This looks good, but should it be applied to the Searches table as well?

keita-determined commented:
> This looks good, but should it be applied to the Searches table as well?

I think so. The ticket specifies run table, and i don't have enough capacities now. Made a ticket https://hpe-aiatscale.atlassian.net/browse/ET-749, so hope someone else can work on it.
